File: cpp/053. Maximum Subarray.cpp
#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int *dp = new int[nums.size()];
        dp[0] = nums[0];
        int max = dp[0];
        for(int i = 1; i < nums.size(); i++){
        	dp[i] = dp[i - 1] > 0 ? (dp[i - 1] + nums[i]) : nums[i];
        	max = max > dp[i] ? max : dp[i];
        }
        return max;
    }
};

File: java/Array/053. Maximum Subarray.java
// DP
public class Solution {
    public int maxSubArray(int[] nums) {
        int max = Integer.MIN_VALUE, cur = 0;

        for(int i: nums) {
            if(cur <= 0) cur = i;
            else cur += i;

            max = cur > max ? cur : max;
        }

        return max;
    }
}
